Source/WebCore/Modules/model-element/HTMLModelElement.cpp

```diff
@@ -698,18 +705,39 @@ void HTMLModelElement::setEnvironmentMap(const URL& url)
 
     m_environmentMapURL = url;
 
-    m_pendingEnvironmentMapData.reset();
+    environmentMapResetAndReject(Exception { ExceptionCode::AbortError });
+    m_environmentMapReadyPromise = makeUniqueRef<EnvironmentMapPromise>();
 
-    if (m_environmentMapResource) {
-        m_environmentMapResource->removeClient(*this);
-        m_environmentMapResource = nullptr;
+    // empty string has been provided, requires environment map removal
+    const auto& attr = attributeWithoutSynchronization(environmentmapAttr).string().trim(isASCIIWhitespace);
+    if (StringView(attr).containsOnly<isASCIIWhitespace<UChar>>()) {
+        // sending a message with empty data to indicate resource removal
+        if (m_modelPlayer)
+            m_modelPlayer->setEnvironmentMap(SharedBuffer::create());
+        return;
     }
```

Comment:
Should this logic of checking for empty string go into `HTMLModelElement::selectEnvironmentMapURL()`? And have that method returns a `std::optional<URL>`? 

`HTMLModelElement::setEnvironmentMap()` can then take in an optional URL too, and remove the environment map if the URL is null?

Reply:
I agree that checking attribute emptiness should be moved to `selectEnvironmentMapURL` but we still need a check `m_environmentMapURL.isEmpty()` in `setEnvironmentMap` to avoid unnecessary resource requests.

As for std::optional<URL> I'd prefer not to use it here because a) for consistency sake (see model source URL), b) `URL()` covers empty case.

Updated code.